### DISCUSSION TITLE
Ensure atomic inside interceptor ?

### DISCUSSION THREAD

**Discussion by Ninja4Panda** [simolus3/drift#2853] [Q&A] [answered]
I want to log every action of a user to the database so I was using an interceptor like so. 
![image](https://github.com/simolus3/drift/assets/61926182/f4f83852-ca71-4d59-aae8-d5216480b620)

To ensure it is atomic, I did this.
![image](https://github.com/simolus3/drift/assets/61926182/595aa77d-7a49-40f3-9a05-34611b871eb3)

Is there any ways to start a transaction within the interceptor so there is no need to add transaction before every query ?

**Comment by simolus3** [comment, marked as answer]
> Is there any ways to start a transaction within the interceptor so there is no need to add transaction before every query ?

I haven't tested this, but you can call `beginTransaction` on the executor to open a transaction executor:

```dart
class _NullUser extends QueryExecutorUser {
  @override
  Future<void> beforeOpen(QueryExecutor executor, OpeningDetails details) {
    throw UnimplementedError();
  }

  @override
  int get schemaVersion => throw UnimplementedError();
}
```

```dart
  @override
  Future<List<Map<String, Object?>>> runSelect(
      QueryExecutor executor, String statement, List<Object?> args) async {
    if (statements.startsWith('SELECT')) {
      return super.runSelect(executor, statement, args);
    }
    final tx = executor.beginTransaction();
    await tx.ensureOpen(_NullUser()); // database connection is already open, so we can use a fake user
    // ... prepare logs
    await tx.runInsert(log, logArgs);
    final result = super.runSelect(tx, statement, args);
    await tx.close();
  }
```

But if you want to keep a transcript of changes, it might be easier to setup [triggers](https://sqlite.org/lang_createtrigger.html) for that.

**Reply by Ninja4Panda** [reply]
I did tried doing it using a transaction executor before but it didn't work. Maybe I did something wrong tho...
As you can see, I even tried using a batch and saw the "[Used internally by drift](https://drift.simonbinder.eu/api/drift/argumentsforbatchedstatement-class)" so I just stopped there
![image](https://github.com/simolus3/drift/assets/61926182/7352f6f4-f9ba-468f-8fce-8cc193bea544)

Anyhow, thanks for the reply. I've moved on to using trigger instead.
